File: taxease-frontend/src/app/core/services/payment.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { shareReplay, tap } from 'rxjs/operators';
import { environment } from '../../../environments/environment';
import { PaymentRequest, PaymentResponse, PaymentMetrics, RevenueDashboard, RevenueRecord, PaymentMethod, PaymentStatus } from '../models/payment.model';

@Injectable({ providedIn: 'root' })
export class PaymentService {
  private base = `${environment.apiUrl}/payments`;
  private cache = new Map<string, Observable<any>>();

  constructor(private http: HttpClient) {}

  private cached<T>(key: string, req: Observable<T>): Observable<T> {
    if (!this.cache.has(key)) this.cache.set(key, req.pipe(shareReplay(1)));
    return this.cache.get(key)!;
  }

  invalidate(): void { this.cache.clear(); }

  makePayment(payload: PaymentRequest): Observable<PaymentResponse> {
    return this.http.post<PaymentResponse>(`${this.base}/pay`, payload).pipe(
      tap(() => this.cache.clear())
    );
  }

  getPaymentById(id: number): Observable<PaymentResponse> {
    return this.cached(`payment-${id}`, this.http.get<PaymentResponse>(`${this.base}/${id}`));
  }

  getPaymentsByFiling(filingId: number): Observable<PaymentResponse[]> {
    return this.cached(`payments-filing-${filingId}`, this.http.get<PaymentResponse[]>(`${this.base}/filing/${filingId}`));
  }

  getPaymentHistory(taxpayerId: number): Observable<PaymentResponse[]> {
    return this.cached(`history-${taxpayerId}`, this.http.get<PaymentResponse[]>(`${this.base}/history/${taxpayerId}`));
  }

  getAllPayments(): Observable<PaymentResponse[]> {
    return this.cached('all-payments', this.http.get<PaymentResponse[]>(this.base));
  }

  updatePaymentStatus(paymentId: number, status: PaymentStatus): Observable<PaymentResponse> {
    const params = new HttpParams().set('status', status);
    return this.http.put<PaymentResponse>(`${this.base}/${paymentId}/status`, {}, { params }).pipe(
      tap(() => this.cache.clear())
    );
  }

  retryPayment(oldPaymentId: number, newMethod: PaymentMethod): Observable<PaymentResponse> {
    const params = new HttpParams().set('newMethod', newMethod);
    return this.http.post<PaymentResponse>(`${this.base}/retry/${oldPaymentId}`, {}, { params }).pipe(
      tap(() => this.cache.clear())
    );
  }

  getPaymentMetrics(): Observable<PaymentMetrics> {
    return this.cached('metrics', this.http.get<PaymentMetrics>(`${this.base}/metrics`));
  }

  getRevenueDashboard(): Observable<RevenueDashboard> {
    return this.cached('revenue', this.http.get<RevenueDashboard>(`${this.base}/revenue`));
  }

  getRevenueByTaxpayer(taxpayerId: number): Observable<RevenueRecord[]> {
    return this.cached(`revenue-tp-${taxpayerId}`, this.http.get<RevenueRecord[]>(`${this.base}/revenue/taxpayer/${taxpayerId}`));
  }
}

File: payment-service/src/main/java/com/paymentservice/service/PaymentService.java
package com.paymentservice.service;

import com.paymentservice.dto.PaymentMetricsResponse;
import com.paymentservice.dto.PaymentRequest;
import com.paymentservice.dto.PaymentResponse;
import com.paymentservice.dto.RevenueDashboardResponse;
import com.paymentservice.dto.RevenueResponse;
import com.paymentservice.entity.enums.PaymentMethod;
import com.paymentservice.entity.enums.PaymentStatus;

import java.util.List;

public interface PaymentService {
    List<PaymentResponse> getAllPayments();
    PaymentResponse initiatePayment(PaymentRequest request);
    PaymentResponse getPaymentById(Long paymentId);
    List<PaymentResponse> getPaymentsByFiling(Long filingId);
    List<PaymentResponse> getPaymentsByTaxpayer(Long taxpayerId);
    PaymentResponse updatePaymentStatus(Long paymentId, PaymentStatus status);
    PaymentResponse retryPayment(Long oldPaymentId, PaymentMethod newMethod);
    PaymentMetricsResponse getPaymentMetrics();
    RevenueDashboardResponse getRevenueDashboard();
    List<RevenueResponse> getRevenueByTaxpayer(Long taxpayerId);
}
